File: app/src/test/java/com/test/casenio/MainViewModelTest.java
package com.test.casenio;

import com.test.casenio.old.MainViewModel;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;

import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class MainViewModelTest {

    @Mock
    private MessageClient mMessageClient;

    private MainViewModel mViewModel;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        setUpSchedulers();
        mViewModel = new MainViewModel(mMessageClient);
    }

    @After
    public void tearDown() {
        RxJavaPlugins.reset();
        RxAndroidPlugins.reset();
    }

    @Test
    public void connectToClient() {
        Completable completable = Completable.complete();
        when(mMessageClient.connect()).thenReturn(completable);
        when(mMessageClient.publish(anyString(), anyString())).thenReturn(completable);
        when(mMessageClient.subscribe(anyString())).thenReturn(Observable.just(""));

        mViewModel.connectToClient();

        verify(mMessageClient).connect();
        verify(mMessageClient).publish(anyString(), anyString());
        verify(mMessageClient).subscribe(anyString());
    }

    @Test
    public void disconnectFromClient() {
        Completable completable = Completable.complete();
        when(mMessageClient.unsubscribe(anyString())).thenReturn(completable);
        when(mMessageClient.disconnect()).thenReturn(completable);

        mViewModel.disconnectFromClient();

        verify(mMessageClient).unsubscribe(anyString());
        verify(mMessageClient).disconnect();
    }

    private void setUpSchedulers() {
        RxJavaPlugins.setIoSchedulerHandler(scheduler -> Schedulers.trampoline());
        RxAndroidPlugins.setInitMainThreadSchedulerHandler(schedulerCallable -> Schedulers.trampoline());
    }

}

File: app/src/test/java/com/test/casenio/MainViewModelTest.kt
package com.test.casenio

import com.test.casenio.messageclient.MessageClient
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.android.plugins.RxAndroidPlugins
import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.schedulers.Schedulers
import org.junit.After
import org.junit.Before
import org.junit.Test

import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations

class MainViewModelTest {

    @Mock
    private lateinit var messageClient: MessageClient
    private lateinit var viewModel: MainViewModel

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        setUpSchedulers()
        viewModel = MainViewModel(messageClient)
    }

    @After
    fun tearDown() {
        RxJavaPlugins.reset()
        RxAndroidPlugins.reset()
    }

    @Test
    fun connectToClient() {
        val completable = Completable.complete()
        `when`<Completable>(messageClient.connect()).thenReturn(completable)
        `when`<Completable>(messageClient.publish(anyString(), anyString()))
                .thenReturn(completable)
        `when`<Observable<String>>(messageClient.subscribe(anyString()))
                .thenReturn(Observable.just(""))

        viewModel.connectToClient()

        verify<MessageClient>(messageClient).connect()
        verify<MessageClient>(messageClient).publish(anyString(), anyString())
        verify<MessageClient>(messageClient).subscribe(anyString())
    }

    @Test
    fun disconnectFromClient() {
        val completable = Completable.complete()
        `when`(messageClient.unsubscribe(anyString())).thenReturn(completable)
        `when`(messageClient.disconnect()).thenReturn(completable)

        viewModel.disconnectFromClient()

        verify(messageClient).unsubscribe(anyString())
        verify(messageClient).disconnect()
    }

    private fun setUpSchedulers() {
        RxJavaPlugins.setIoSchedulerHandler { Schedulers.trampoline() }
        RxAndroidPlugins.setInitMainThreadSchedulerHandler { Schedulers.trampoline() }
    }
}